### PR TITLE
Add Cadence Program Pragmas

### LINKED ISSUE #165: Version programs
## Context

Cadence programs should indicate in what version of Cadence they are written in.

## Definition of Done

- [ ] Parser: 
  - [ ] Require the static annotation `#version(_ version: String)`, e.g. `#version("0.5")` at the beginning of programs
  - [ ] AST: Add `version string` field to `ast.Program`
- [ ] Tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by benjaminkvm**
Closes #165 

## Description

This PR adds support for Cadence program pragmas. 

Pragmas are internally represented as top-level declarations containing either a `IdentifierExpression` (`#pedantic`) or a `InvocationExpression` (`#version("1.0")`), preceded by a `#` token/sigil. Pragma declarations are available within a `Program` via `program.PragmaDeclarations`.

Small refactor to support `ParseAndCheckWithOptions` with `OnlyNewParser`, for the pragma test cases.

Current Limitations:
- Only available in the the new parser.
- Pragmas do not support type arguments and non-`ast.StringExpresssion` arguments.

### CODE REVIEW

File: runtime/sema/check_pragma.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package sema
+
+import "github.com/onflow/cadence/runtime/ast"
+
+func (checker *Checker) VisitPragmaDeclaration(p *ast.PragmaDeclaration) ast.Repr {
+
+	invocPragma, isInvocPragma := p.Expression.(*ast.InvocationExpression)
+	_, isIdentPragma := p.Expression.(*ast.IdentifierExpression)
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:26]: Slight optimization: Maybe only check if the expression is an identifier expression if it is  not known to be an invocation expression

```suggestion
	var isIdentPragma bool
	if !isInvocPragma {
		_, isIdentPragma = p.Expression.(*ast.IdentifierExpression)
	}
```

File: runtime/sema/check_pragma.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	// Pragma can be either an invocation expression or an identfier expression
+	if !(isInvocPragma || isIdentPragma) {
+		checker.report(&InvalidPragmaError{Message: "must be identifier or invocation expression"})
+	}
+
+	if isInvocPragma {
+		// Type arguments are not supported for pragmas
+		if invocPragma.TypeArguments != nil {
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:35]: Maybe check length here, so the test is not accidentally true if `invocPragma.TypeArguments` is an allocated empty slice

File: runtime/sema/check_pragma.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import "github.com/onflow/cadence/runtime/ast"
+
+func (checker *Checker) VisitPragmaDeclaration(p *ast.PragmaDeclaration) ast.Repr {
+
+	invocPragma, isInvocPragma := p.Expression.(*ast.InvocationExpression)
+	_, isIdentPragma := p.Expression.(*ast.IdentifierExpression)
+
+	// Pragma can be either an invocation expression or an identfier expression
+	if !(isInvocPragma || isIdentPragma) {
+		checker.report(&InvalidPragmaError{Message: "must be identifier or invocation expression"})
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:30]: `InvalidPragmaError` needs position information. Maybe use the range of `p.Expression`

File: runtime/sema/check_pragma.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	if isInvocPragma {
+		// Type arguments are not supported for pragmas
+		if invocPragma.TypeArguments != nil {
+			checker.report(&InvalidPragmaError{Message: "type arguments not supported"})
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:36]: `InvalidPragmaError` needs position information

File: runtime/sema/check_pragma.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+		}
+		// Ensure arguments are string expressions
+		for _, arg := range invocPragma.Arguments {
+			_, ok := arg.Expression.(*ast.StringExpression)
+			if !ok {
+				checker.report(&InvalidPragmaError{Message: "invalid arguments"})
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:42]: `InvalidPragmaError` needs position information

File: runtime/sema/errors.go [hunk 6]
```diff
@@ -53,6 +53,16 @@ func (e *unsupportedOperation) Error() string {
 	)
 }
 
+// InvalidPragmaError
+
+type InvalidPragmaError struct {
+	Message string
+}
+
```
Comment (turbolent) [inline runtime/sema/errors.go:61]: As mentioned above, this error needs position information. Maybe just embed `ast.Range` like in the other errors. Also, the error should conform to `SemanticError` (done by just implementing (`isSemanticError()`):


```suggestion
type InvalidPragmaError struct {
	Message string
	ast.Range
}

func (*InvalidPragmaError) isSemanticError() {}
```

File: runtime/tests/parser/parser_test.go [hunk 7]
```diff
@@ -7162,13 +7227,13 @@ func TestParseTransactionDeclaration(t *testing.T) {
 	t.Run("SimpleTransaction", func(t *testing.T) {
 		const code = `
 		  transaction {
-
```
Comment (turbolent) [inline runtime/tests/parser/parser_test.go:7165]: Thank you for cleaning this up! 🏅

**Comment by benjaminkvm** [comment]
Thanks for the thorough review @turbolent! Should be all good to go!

File: runtime/sema/check_pragma.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+			_, ok := arg.Expression.(*ast.StringExpression)
+			if !ok {
+				checker.report(&InvalidPragmaError{
+					Message: "invalid argument",
+					Range: ast.Range{
+						StartPos: arg.Expression.StartPosition(),
+						EndPos:   arg.Expression.EndPosition(),
+					},
+				})
+				return nil
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:64]: Missed this earlier: We should report as much as we can, so continue checking the remaining arguments and don't return
```suggestion
```

File: runtime/sema/errors.go [hunk 9]
```diff
@@ -53,6 +53,27 @@ func (e *unsupportedOperation) Error() string {
...
+func (e *InvalidPragmaError) isSemanticError() {}
+
+func (e *InvalidPragmaError) StartPosition() ast.Position {
+	return e.Range.StartPos
+}
+
+func (e *InvalidPragmaError) EndPosition() ast.Position {
+	return e.Range.EndPos
+}
```
Comment (turbolent) [inline runtime/sema/errors.go:71]: These are not needed, embedding `ast.Range` makes `InvalidPragmaError` implement the functions already 🎉 

```suggestion
```

File: runtime/sema/check_pragma.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	// Pragma can be either an invocation expression or an identfier expression
+	if !(isInvocPragma || isIdentPragma) {
+		checker.report(&InvalidPragmaError{
+			Message: "must be identifier or invocation expression",
+			Range: ast.Range{
+				StartPos: p.Expression.StartPosition(),
+				EndPos:   p.Expression.EndPosition(),
+			},
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:38]: Can use the existing helper here:

```suggestion
			Range: ast.NewRangeFromPositioned(p.Expression),
```

File: runtime/sema/check_pragma.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+	if isInvocPragma {
+		// Type arguments are not supported for pragmas
+		if len(invocPragma.TypeArguments) > 0 {
+			checker.report(&InvalidPragmaError{
+				Message: "type arguments not supported",
+				Range: ast.Range{
+					StartPos: invocPragma.StartPosition(),
+					EndPos:   invocPragma.EndPosition(),
+				},
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:50]:
```suggestion
				Range: ast.NewRangeFromPositioned(invocPragma),
```

File: runtime/sema/check_pragma.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+		// Ensure arguments are string expressions
+		for _, arg := range invocPragma.Arguments {
+			_, ok := arg.Expression.(*ast.StringExpression)
+			if !ok {
+				checker.report(&InvalidPragmaError{
+					Message: "invalid argument",
+					Range: ast.Range{
+						StartPos: arg.Expression.StartPosition(),
+						EndPos:   arg.Expression.EndPosition(),
+					},
```
Comment (turbolent) [inline runtime/sema/check_pragma.go:62]:
```suggestion
					Range: ast.NewRangeFromPositioned(arg.Expression),
```

**Pull request merged**